pkg/server/tls.go

```diff
@@ -0,0 +1,85 @@
+package server
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	clientCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_client_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC client TLS certificate in seconds since the Unix epoch.",
+	})
+	serverCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_server_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC server TLS certificate in seconds since the Unix epoch.",
+	})
+)
+
+type lazyGauge struct {
+	init   sync.Once
+	metric prometheus.Gauge
+}
+
+func newLazyGauge(opts prometheus.GaugeOpts) *lazyGauge {
+	return &lazyGauge{metric: prometheus.NewGauge(opts)}
+}
+
+func (g *lazyGauge) Set(v float64) {
+	g.init.Do(func() { prometheus.MustRegister(g.metric) })
+	g.metric.Set(v)
+}
+
+func loadVerifiedCerts(certFile, keyFile, caFile string, usage x509.ExtKeyUsage) (cert tls.Certificate, caPool *x509.CertPool, expiry time.Time, err error) {
+	cert, caPool, err = loadCerts(certFile, keyFile, caFile)
+	if err != nil {
+		return tls.Certificate{}, nil, time.Time{}, err
+	}
+	chains, err := verify(cert, caPool, usage)
+	if err != nil {
+		return tls.Certificate{}, nil, time.Time{}, err
+	}
+	return cert, caPool, earliestExpiry(chains), nil
+}
+
+func loadCerts(certFile, keyFile, caFile string) (tls.Certificate, *x509.CertPool, error) {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return tls.Certificate{}, nil, err
+	}
+	ca, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return tls.Certificate{}, nil, err
+	}
+	caPool := x509.NewCertPool()
+	caPool.AppendCertsFromPEM(ca)
+	return cert, caPool, nil
+}
+
+func verify(cert tls.Certificate, pool *x509.CertPool, usage x509.ExtKeyUsage) (chains [][]*x509.Certificate, err error) {
+	x509Cert := cert.Leaf
+	if x509Cert == nil {
+		x509Cert, err = x509.ParseCertificate(cert.Certificate[0])
+	}
+	return x509Cert.Verify(x509.VerifyOptions{
+		Roots:     pool,
+		KeyUsages: []x509.ExtKeyUsage{usage},
+	})
+}
```

Comment:
we didn't do any key usage verification previously so this would potentially be a breaking change for users, so this would either need to be optionally enabled or it would have to a wait for a major version change.

Reply:
No longer required, but tracked in a separate metric.